### PR TITLE
ADBDEV-6338: Implement MVP to paralellize gprestore

### PR DESCRIPTION
**Pull request by RekGRpth**
Make gprestore --resize-cluster use --jobs for parallel restoration

Users consider the current behavior to be incorrect, since the --jobs parameter
is used when creating a backup, which is not actually used during restoration.
Restoration works in a single instance of the gpbackup_helper agent (in the
--restore-agent mode).

In short, when restoring a backup using the gprestore utility with the
--resize-cluster option, which was previously made in parallel creation mode
(--jobs), the utility does not use the corresponding mode during restoration.
In this case, restoration occurs through gpbackup_helper, which is launched in
one instance of this process for each segment. Thus, one table is processed at
a time based on the list of table OIDs passed when gpbackup_helper is launched.

As a solution to the problem, this patch implements the launch of several
gpbackup_helper instances based on the value passed in the --jobs argument.

To do this:

1) The general list of table OIDs for recovery is distributed between several
instances of gpbackup_helper processes according to the specified value of the
--jobs parameter.

2) Since the general list for restoration is sorted by size, the distribution
among gpbackup_helper instances is done according to the principle of
sequentially distributing one table OID to each instance. Thus, the load is
expected to be more uniform. All batches corresponding to the same table OID go
to the same gpbackup_helper instance.

3) An integer identifier of the "instance number" is added to the file name of
the list of table OIDs for restoration by gpbackup_helper instances. They are
not logically related in any way, the main thing is that its own list is passed
as an argument when starting gpbackup_helper.

4) The instance number is also added to the name of the script file
(scriptFile).

5) The name of the pipeFile file remains unchanged.

6) To create the initial list of pipes, one pipe is created for the first table
OID (according to the sorting of table OIDs by decreasing table size), which
will be processed by each instance of gpbackup_helper. Thus, at the time of
starting each instance of gpbackup_helper, the file system will already have
the first pipe corresponding to the first table OID for this instance.

7) When starting gpbackup_helper, the value of the --jobs parameter is taken
into account (in fact, the number of connections in the pool). The number of
launched instances corresponds to the value of the parameter. The value of the
path of the first pipe (pipeFile), the path to the agent startup script
(scriptFile), and the path to the list of table OIDs corresponding to this
instance are passed as arguments.

8) When deleting auxiliary files in DoCleanup, the files with the list of table
OIDs and the startup script files (scriptFile) are deleted.

9) For --single-data-file and --copy-queue-size modes, the current behavior
remains unchanged.

---
It is easier to view the changes with the "Hide whitespace" option enabled.

### CODE REVIEW

**Comment by silent-observer** [comment]
> list of OID tables

Maybe "list of table OIDs"? This is used multiple times in the description.

The code itself LGTM.

**Comment by RekGRpth** [comment]
> Maybe "list of table OIDs"? This is used multiple times in the description.

renamed

**Comment by silent-observer** [comment]
Missed a few "OID table" in 2 and 6

**Comment by RekGRpth** [comment]
> Missed a few "OID table" in 2 and 6

fixed

**Review by silent-observer — APPROVED** [review]

File: utils/agent_remote.go [hunk 1]
```diff
@@ -143,7 +143,7 @@ func VerifyHelperVersionOnSegments(version string, c *cluster.Cluster) {
 	}
 }
 
-func StartGpbackupHelpers(c *cluster.Cluster, fpInfo filepath.FilePathInfo, operation string, pluginConfigFile string, compressStr string, onErrorContinue bool, isFilter bool, wasTerminated *bool, copyQueue int, isSingleDataFile bool, resizeCluster bool, origSize int, destSize int, verbosity int) {
+func StartGpbackupHelpers(c *cluster.Cluster, fpInfo filepath.FilePathInfo, operation string, pluginConfigFile string, compressStr string, onErrorContinue bool, isFilter bool, wasTerminated *bool, copyQueue int, isSingleDataFile bool, resizeCluster bool, origSize int, destSize int, verbosity int, j int) {
```
Comment (whitehawk) [inline utils/agent_remote.go:146]: `j` is not informative name for a function argument, let's rename it.

Comment (whitehawk) [inline utils/agent_remote.go:146]: Number of arguments for this function is indecently large, what do you think about packing them in a structure?

File: backup/backup.go [hunk 2]
```diff
@@ -306,15 +306,15 @@ func backupData(tables []Table) {
 		for _, table := range tables {
 			oidList = append(oidList, fmt.Sprintf("%d", table.Oid))
 		}
-		utils.WriteOidListToSegments(oidList, globalCluster, globalFPInfo, "oid")
+		utils.WriteOidListToSegments(oidList, globalCluster, globalFPInfo, fmt.Sprintf("oid_%d", 0))
```
Comment (whitehawk) [inline backup/backup.go:309]:
```suggestion
		utils.WriteOidListToSegments(oidList, globalCluster, globalFPInfo, "oid_0")
```

File: restore/data.go [hunk 3]
```diff
@@ -215,34 +215,49 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
...
+		} else {
+			maxHelpers = totalTables
+		}
+
+		for j := 0; j < maxHelpers; j++ {
+			// During a larger-to-smaller restore, we need to do multiple passes of
+			// data loading so we assign the batches here.
+			oidList := make([]string, 0)
+			for i, entry := range dataEntries {
+				if i%maxHelpers != j {
```
Comment (whitehawk) [inline restore/data.go:232]:
```suggestion
				if i % maxHelpers != j {
```

File: restore/data.go [hunk 4]
```diff
@@ -215,34 +215,49 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
...
+		var maxHelpers int
+		if backupConfig.SingleDataFile {
+			maxHelpers = 1
+		} else if connectionPool.NumConns < totalTables {
+			maxHelpers = connectionPool.NumConns
+		} else {
+			maxHelpers = totalTables
+		}
+
+		for j := 0; j < maxHelpers; j++ {
```
Comment (whitehawk) [inline restore/data.go:227]: Nice to have: make `j` more informative, for instance, `helperIdx`.

File: restore/data.go [hunk 5]
```diff
@@ -215,34 +215,49 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
...
+			// During a larger-to-smaller restore, we need to do multiple passes of
+			// data loading so we assign the batches here.
+			oidList := make([]string, 0)
+			for i, entry := range dataEntries {
```
Comment (whitehawk) [inline restore/data.go:231]: Nice to have: make `i` more informative, for instance, `entryIdx`.

File: backup/backup.go (hunk 2 above)
Comment (RekGRpth) [reply backup/backup.go:309]: applied

File: restore/data.go (hunk 4 above)
Comment (RekGRpth) [reply restore/data.go:227]: renamed

File: restore/data.go (hunk 5 above)
Comment (RekGRpth) [reply restore/data.go:231]: renamed

File: restore/data.go (hunk 3 above)
Comment (RekGRpth) [reply restore/data.go:232]: VSCode formats the file this way when saving it.

File: utils/agent_remote.go (hunk 1 above)
Comment (RekGRpth) [reply utils/agent_remote.go:146]: renamed

Comment (RekGRpth) [reply utils/agent_remote.go:146]: it's MVP only, not refactoring

File: restore/data.go (hunk 3 above)
Comment (RekGRpth) [reply restore/data.go]: formatted via additional braces

**Comment by whitehawk** [comment]
gpbackup_helper in all logs uses prefix "Segment %d: ". It was enough to distinguish logs from different helpers while there was only 1 helper per segment.
But now it is possible to have many instances of the helper on the same segment. We need to update logging to be able to distinguish logs from different instances.

**Comment by whitehawk** [comment]
Multiple instances of gpbackup_helper still share the same filename for the error file. So, if one instance fails, it could affect the work of others (for ex. the errorFile is checked in the copy command created in `CopyTableIn()`)
Did you test such scenario?

**Comment by RekGRpth** [comment]
> gpbackup_helper in all logs uses prefix "Segment %d: ". It was enough to distinguish logs from different helpers while there was only 1 helper per segment. But now it is possible to have many instances of the helper on the same segment. We need to update logging to be able to distinguish logs from different instances.

Each helper writes its own PID to the log, so now on one segment they can be distinguished by PIDs.

**Comment by RekGRpth** [comment]
> Multiple instances of gpbackup_helper still share the same filename for the error file. So, if one instance fails, it could affect the work of others (for ex. the errorFile is checked in the copy command created in `CopyTableIn()`) Did you test such scenario?

This is the expected behavior that after creating the error file all work should be completed, and this is already checked in the test "Will not hang after error during restore with jobs".

**Comment by whitehawk** [comment]
> > Multiple instances of gpbackup_helper still share the same filename for the error file. So, if one instance fails, it could affect the work of others (for ex. the errorFile is checked in the copy command created in `CopyTableIn()`) Did you test such scenario?
> 
> This is the expected behavior that after creating the error file all work should be completed, and this is already checked in the test "Will not hang after error during restore with jobs".

hm, what do you think about possibility of a following situation:
1. Let's say we have 2 jobs to restore 2 tables. 1st table is significantly smaller then 2nd (thus will complete faster). And backup data for each table contains an error.
2. gprestore starts 2 instances of the helper on each segment. 1st helper serves 1st table, and 2nd helper serves 2nd table.
3. gprestore also starts 2 goroutins, where `restoreSingleTableData()` is called.
4. 1st helper detects an error and creates an error file.
5. 1st goroutine detects the error file by `CheckAgentErrorsOnSegments()` (called from `restoreSingleTableData()`). `CheckAgentErrorsOnSegments()` deletes the error file created by 1st helper. And gprestore stops execution because of consequent FatalOnError.
6. But the second helper is not stopped, right? 2nd helper detects an error and creates its own error file, which persists in the file system till next restore...

**Comment by RekGRpth** [comment]
> 6\. But the second helper is not stopped, right?

Why? I think it will be stopped (killed) at the cleanup. And also during cleanup all auxiliary files will be deleted.

**Comment by whitehawk** [comment]
> > 6. But the second helper is not stopped, right?
> 
> Why? I think it will be stopped (killed) at the cleanup. And also during cleanup all auxiliary files will be deleted.

It looks that `CleanUpHelperFilesOnAllHosts()` doesn't cleanup "_error" file.
`CheckAgentErrorsOnSegments()` does cleanup of "_error" file, but it is called in `DoCleanup()` only if `wasTerminated` is `true`. But that could be only if gprestore was terminated by SIGINT or SIGTERM. So it looks not our case...

Thus, there is a very tight time window between "goroutine detects the error file by `CheckAgentErrorsOnSegments()` (called from `restoreSingleTableData()`). `CheckAgentErrorsOnSegments()` deletes the error file created by 1st helper" and `DoCleanup()` calls `CleanUpSegmentHelperProcesses()`.
If 2nd helper manages to create an "_error" file in this short time frame, we could get a flaky problem with very low reproduction rate.

Hm, and BTW, if the helper was terminated by `CleanUpSegmentHelperProcesses()` (from `DoCleanup()`), it will anyway generate an [error file](https://github.com/arenadata/gpbackup/blob/b164a2782fbd54eae9b490aa284de1d24916961c/helper/helper.go#L277)...

**Comment by RekGRpth** [comment]
> `CleanUpHelperFilesOnAllHosts`

And the problem is that after completion an error file may remain?

**Comment by whitehawk** [comment]
> And the problem is that after completion an error file may remain?

> Maybe this is a separate task, not related to this patch? After all, we can just kill the helper and it will create an error file that will not be deleted.

Yes, the problem is that after the completion the error file remains. I'm worried that it will spoil the consequent attempt to restore.

**Review by whitehawk — APPROVED** [review]

**Comment by RekGRpth** [comment]
> I'm worried that it will spoil the consequent attempt to restore.

Error filename contains PID of main process.

**Review by silent-observer — APPROVED** [review]

**Comment by RekGRpth** [comment]
> Yes, the problem is that after the completion the error file remains

Still, with and without the patch, the error file is also deleted by the `CleanUpHelperFilesOnAllHosts` function, because it satisfies the regular expression in the `findCommandStr` function.
```
find /home/gpadmin/.data/6/dbfast1/demoDataDir0 -regextype posix-extended -regex ".*gpbackup_0_20240812201233_(oid|script|pipe)_329650.*"
/home/gpadmin/.data/6/dbfast1/demoDataDir0/gpbackup_0_20240812201233_oid_329650
/home/gpadmin/.data/6/dbfast1/demoDataDir0/gpbackup_0_20240812201233_script_329650
/home/gpadmin/.data/6/dbfast1/demoDataDir0/gpbackup_0_20240812201233_pipe_329650_error
```

**Comment by RekGRpth** [comment]
https://github.com/arenadata/gpbackup/pull/110

**Pull request closed without merging**